### PR TITLE
Prepare transition steps

### PR DESCRIPTION
**Pull request by deepthidevaki**
## Description

In this PR, we implement two Steps using the new interface `PartitionTransitionStep`. In addition some more necessary field are added to the `PartitionTransitionContext`. They are needed for some transition steps.

Note that the new transition steps are not yet used. The new transition logic will be used when all steps are moved.

## Related issues

Related #7515 

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [x] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/zeebe-io/zeebe/compare/stable/0.24...develop?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/0.25`) to the PR, in case that fails you need to create backports manually.

Testing:
* [ ] There are unit/integration tests that verify all acceptance criterias of the issue
* [ ] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark 

Documentation: 
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] New content is added to the [release announcement](https://drive.google.com/drive/u/0/folders/1DTIeswnEEq-NggJ25rm2BsDjcCQpDape)

### CODE REVIEW

**Comment by deepthidevaki** [comment]
@pihme This is mostly what we did together last week. I cleaned it up a little and tried to split it up into multiple independent commits.

**Review by pihme — APPROVED** [review]
lgtm :tada: 

Only wish I would have is to add more unit tests for the steps. 
See e.g. https://github.com/camunda-cloud/zeebe/pull/7801/files#diff-5cb60825d3938dcc68caa71031f726b50674c4578db919cf05e20a98035290fe for an example. This heavily uses Mockito, so there might be objections on that ground alone. However, writing the tests I did find some bugs in my implementation.

Anyway, feel free to ignore this comment.

**Comment by deepthidevaki** [comment]
@pihme I have now added tests for both StreamProcessor and zeebedb steps. I had to change `StreamProcessorTransitionStep` to make it testable. Please have a look at it. I will wait for your review before merging this PR.

**Comment by pihme** [comment]
@deepthidevaki Will need a bit more time for the review. I like the tests on first glance. But so far I only looked at it causally.

First impressions:
I wish we could do without the `transitionTo(..)` in the when clause of the tests. But I have to think harder about it.
In my tests I had to make sure to clean up the resources that were created during the test. Some of them bound to ports etc. I was surprised that e.g. `ZeebeDbPartitionTransitionStepTest` didn't have a teardown with `if stateController!= null then stateController.closeDB()` or something along those lines.

As I said, I will do a more thorough review soon.

**Comment by deepthidevaki** [comment]
Thanks @pihme I will check how to do a proper tear down in these tests.

**Comment by deepthidevaki** [comment]
@pihme Actually teardown is not needed as no services are opened. The tests are using mocks for both StreamProcessor and ZeebeDb.

File: broker/src/test/java/io/camunda/zeebe/broker/system/partitions/impl/steps/TestPartitionTransitionContext.java
Comment (pihme) [inline broker/src/test/java/io/camunda/zeebe/broker/system/partitions/impl/steps/TestPartitionTransitionContext.java:32]: Please move to `io.camunda.zeebe.broker.system.partitions`

File: broker/src/test/java/io/camunda/zeebe/broker/system/partitions/impl/steps/TestPartitionTransitionContext.java
Comment (pihme) [inline broker/src/test/java/io/camunda/zeebe/broker/system/partitions/impl/steps/TestPartitionTransitionContext.java:44]:
```suggestion
```
Field is never used

File: broker/src/test/java/io/camunda/zeebe/broker/system/partitions/impl/steps/TestPartitionTransitionContext.java
Comment (pihme) [inline broker/src/test/java/io/camunda/zeebe/broker/system/partitions/impl/steps/TestPartitionTransitionContext.java:34]: :wrench: (optional)
```suggestion
  private RaftPartition raftPartition = Mockito.mock(RaftPartition.class);
```
Then we have a default implementation. And tests could do:

```
when(testPartitionContext.getRaftPartition().getRole()).thenReturn(Role.FOLLOWER);
```

Later we could add a setter to overwrite the default implementation.
Same suggestion for `exporterDirector`

File: broker/src/test/java/io/camunda/zeebe/broker/system/partitions/impl/steps/TestPartitionTransitionContext.java
Comment (pihme) [inline broker/src/test/java/io/camunda/zeebe/broker/system/partitions/impl/steps/TestPartitionTransitionContext.java:59]:
```suggestion
    return TestActorFuture.completedFuture(null);
```
same for `...OfBecomingFollower`

File: broker/src/test/java/io/camunda/zeebe/broker/system/partitions/impl/steps/TestPartitionTransitionContext.java
Comment (pihme) [inline broker/src/test/java/io/camunda/zeebe/broker/system/partitions/impl/steps/TestPartitionTransitionContext.java:97]:
```suggestion
    return true;
```

File: broker/src/test/java/io/camunda/zeebe/broker/system/partitions/impl/steps/ZeebeDbPartitionTransitionStepTest.java
Comment (pihme) [inline broker/src/test/java/io/camunda/zeebe/broker/system/partitions/impl/steps/ZeebeDbPartitionTransitionStepTest.java:108]: Would this also apply to `null -> Role.INACTIVE` ?

File: broker/src/test/java/io/camunda/zeebe/broker/system/partitions/impl/steps/ZeebeDbPartitionTransitionStepTest.java
Comment (pihme) [inline broker/src/test/java/io/camunda/zeebe/broker/system/partitions/impl/steps/ZeebeDbPartitionTransitionStepTest.java:116]: I think this should also happen on `null -> Role.CANDIDATE`

I guess this is not possible in a Raft state machine. But we could have a weird timing in which we missed the transition to follower, or an error occurred on that transition. And then we would see a direct transition from null to candidate.

Thinking more about it, this might be a broader issue. Right now the tests cover transitions as per the Raft state machine. But we could also see transitions not possible in the Raft state machine.

For example:

1. Raft transition: CANDIDATE -> LEADER / Zeebe transition from FOLLOWER to LEADER succeeds
2. Raft transition: LEADER -> FOLLOWER / Zeebe transition from LEADER to FOLLOWER has an error, therefore `setCurrentRole` is never called. In this case the Zeebe partition would still report `currentRole == LEADER` becasue it was the last successful transition
3. Raft transition FOLLOWER-> CANDIDATE / Zeebe transition from LEADER to CANDIDATE 

It's maybe a bad example, because CANDIDATE doesn't mean much for us. But what I do want to show is that more transitions are possible for the steps then what Raft state model allows.

File: broker/src/test/java/io/camunda/zeebe/broker/system/partitions/impl/steps/StreamProcessorTransitionStepTest.java
Comment (pihme) [inline broker/src/test/java/io/camunda/zeebe/broker/system/partitions/impl/steps/StreamProcessorTransitionStepTest.java:132]: See comment on other test regarding transitions not possible in Raft.

Also missing in this test (from Raft state machine):
* null -> INACTIVE
* FOLLOWER -> INACTIVE
* LEADER -> INACTIVE

File: broker/src/test/java/io/camunda/zeebe/broker/system/partitions/impl/steps/StreamProcessorTransitionStepTest.java
Comment (pihme) [inline broker/src/test/java/io/camunda/zeebe/broker/system/partitions/impl/steps/StreamProcessorTransitionStepTest.java:91]: :thought_balloon: One thing I do not understand is when `clearInvocations(streamProcessor)` is needed and when not.

I was wandering why it's not in setup, but maybe this doesn't work with parameterized tests. Is it not needed here, because it is never invoked in one invocation of the parameterized test?

Or could it maybe be needed in the future, when there are other test methods which have interactions. And then this test could fail, because it doesn't clear the invocations.

Maybe it would be easier to understand if `clearInvocations(streamProcessor)` was at the end of a then block which did verify invocations. Then this would better fit the pattern that the test cleans up after itself.

Just a thought, though. Would love to hear your opinions on this and why you did it this way.

**Review by pihme — CHANGES_REQUESTED** [review]
Love the tests! Thank you so much for adding them.

However, some transitions are not yet covered.

File: broker/src/test/java/io/camunda/zeebe/broker/system/partitions/impl/steps/StreamProcessorTransitionStepTest.java
Comment (deepthidevaki) [reply broker/src/test/java/io/camunda/zeebe/broker/system/partitions/impl/steps/StreamProcessorTransitionStepTest.java:91]: In test like the following
```
   // given
    if (currentRole != null) {
      transitionTo(currentRole);
    }
    Mockito.clearInvocations(streamProcessor);

    // when
    transitionTo(targetRole);

    // then  
    assertThat(transitionContext.getStreamProcessor()).isNotNull();
    verify(streamProcessor).openAsync(anyBoolean());
```
The `transitionTo` in "given" interacts with mock. `clearInvocations` is to clear those interactions so that we can then test the expected in "then". Alternatively, what I did for the tests for the rest of the steps which is not yet in this PR is as follows:
```
   // given
    context.setCurrentRole(currentRole)
    if (currentRole != null) {
      context.setStreamProcessor(mockStreamProcessorFromPrevRole);
    }

    // when
    transitionTo(targetRole);

    // then  
    assertThat(transitionContext.getStreamProcessor()).isNotNull().isNotEqualTo(mockStreamProcessorFromPrevRole);
    verify(streamProcessor).openAsync(anyBoolean());
```

In this way there is no real steps called in "given" and thus no invocations to the mock. If this is more clear, I will change the tests to follow this pattern. Wdyt?

Comment (pihme) [reply broker/src/test/java/io/camunda/zeebe/broker/system/partitions/impl/steps/StreamProcessorTransitionStepTest.java:91]: Love it!

File: broker/src/test/java/io/camunda/zeebe/broker/system/partitions/impl/steps/ZeebeDbPartitionTransitionStepTest.java
Comment (deepthidevaki) [reply broker/src/test/java/io/camunda/zeebe/broker/system/partitions/impl/steps/ZeebeDbPartitionTransitionStepTest.java:116]: Errors in installation are a bit more problematic, I think. If steps are partially installed there might be some inconsistency. If LEADER -> FOLLOWER transition failed in Zeebe, I think we should never do "Zeebe transition from LEADER to CANDIDATE". If a transition fails in between, we should probably close all, may be transition to INACTIVE to ensure that we restart with a clean state. But that is a different topic. Here for the sake of completeness, I will add tests for the all possible transitions. Although I have to think what should be the semantics of those.

File: broker/src/test/java/io/camunda/zeebe/broker/system/partitions/impl/steps/StreamProcessorTransitionStepTest.java
Comment (deepthidevaki) [reply broker/src/test/java/io/camunda/zeebe/broker/system/partitions/impl/steps/StreamProcessorTransitionStepTest.java:155]: There is a specific test for transition to Inactive `shouldCloseStreamProcessorWhenTransitioningToInactive`.

**Comment by deepthidevaki** [comment]
@pihme I have updated the tests and applied your other comments. Please take a look.

**Review by pihme — APPROVED** [review]
lgtm :tada:

**Pull request merged**
